### PR TITLE
refactor: buffer, edge, bucket

### PR DESCRIPTION
**Pull request by whynowy**
1. Refactor according to https://github.com/numaproj/numaflow/blob/main/docs/specifications/edges-buffers-buckets.md.
2. Deprecate `spec.edges[].parallelism`, use `spec.vertices[].partitions`.
3. Deprecate `spec.edges[].limits.*`, use `spec.vertices[].limits.*`, `*` includes `bufferMaxLength` and `bufferUsageLimit`.
4. TODO: UI is broken, need to fix.
5. TODO: existing buffer/bucket renaming to make it backward compatible - this does not seem to be doable, because JetStream does not support renaming, and there's no way to create a new stream and consumer, meanwhile copy the same data and keep the same offsets and same consumer ack record.

### CODE REVIEW

**Comment by vigith** [comment]
> 5. TODO: existing buffer/bucket renaming to make it backward compatible - this does not seem to be doable, because JetStream does not support renaming, and there's no way to create a new stream and consumer, meanwhile copy the same data and keep the same offsets and same consumer ack record.

Do we really care whether this has be to backward compatible?

**Comment by whynowy** [comment]
> > 5. TODO: existing buffer/bucket renaming to make it backward compatible - this does not seem to be doable, because JetStream does not support renaming, and there's no way to create a new stream and consumer, meanwhile copy the same data and keep the same offsets and same consumer ack record.
> 
> Do we really care whether this has be to backward compatible?

It would be nice to have, but it seems no way to do it...

**Review by vigith — COMMENTED** [review]
LGTM, I did not get the complete picture of why we introduced the `CombinedEdge` struct. Could you please add some comments on that struct?

File: pkg/apis/numaflow/v1alpha1/edge_types.go [hunk 1]
```diff
@@ -38,6 +44,40 @@ type Edge struct {
...
+	// The number of partitions of the from vertex, if not provided, the default value is set to "1".
+	// +optional
+	FromVertexPartitions *int32 `json:"fromVertexPartitions,omitempty" protobuf:"bytes,3,opt,name=fromVertexPartitions"`
+	// +optional
+	FromVertexLimits *VertexLimits `json:"fromVertexLimits,omitempty" protobuf:"bytes,4,opt,name=fromVertexLimits"`
+	// To vertex type.
+	ToVertexType VertexType `json:"toVertexType" protobuf:"bytes,5,opt,name=toVertexType"`
+	// The number of partitions of the to vertex, if not provided, the default value is set to "1".
+	// +optional
+	ToVertexPartitions *int32 `json:"toVertexPartitions,omitempty" protobuf:"bytes,6,opt,name=toVertexPartitions"`
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/apis/numaflow/v1alpha1/edge_types.go:60]: This is right, we need to know what are the write buffers in a vertex pod.

File: pkg/apis/numaflow/v1alpha1/edge_types.go [hunk 2]
```diff
@@ -38,6 +44,40 @@ type Edge struct {
...
+	// +optional
+	ToVertexLimits *VertexLimits `json:"toVertexLimits,omitempty" protobuf:"bytes,7,opt,name=toVertexLimits"`
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/apis/numaflow/v1alpha1/edge_types.go:62]: `FromVertexLimits` is not used now, adding it is just for symmetry.

File: pkg/apis/numaflow/v1alpha1/edge_types.go [hunk 3]
```diff
@@ -38,6 +44,40 @@ type Edge struct {
 	OnFull *BufferFullWritingStrategy `json:"onFull,omitempty" protobuf:"bytes,6,opt,name=onFull"`
 }
 
+type CombinedEdge struct {
+	Edge `json:",inline" protobuf:"bytes,1,opt,name=edge"`
+	// From vertex type.
+	FromVertexType VertexType `json:"fromVertexType" protobuf:"bytes,2,opt,name=fromVertexType"`
+	// The number of partitions of the from vertex, if not provided, the default value is set to "1".
+	// +optional
+	FromVertexPartitions *int32 `json:"fromVertexPartitions,omitempty" protobuf:"bytes,3,opt,name=fromVertexPartitions"`
```
[the remark this answers is not in the document]
Comment (whynowy) [reply pkg/apis/numaflow/v1alpha1/edge_types.go:53]: `FromVertexPartitioins` won't be used in the future, adding it for 2 reasons:

1. Symmetry
2. It's now used for reduce vertex to know what are the indexed reading buckets.

**Review by vigith — APPROVED** [review]

**Pull request merged**